mountpoint-s3/src/prefetch/parquet_stream.rs

```diff
@@ -0,0 +1,423 @@
+    if metadata.is_none() {
+        return false;
+    }
+
+    let rowgroup_col_cache = in_mem_cache.read().await;
+    if let Some(cache) = &*rowgroup_col_cache {
+        for (row_group_col, col_range) in rowgroup_cols {
+            if let Some(col_cache) = cache.get(row_group_col) {
+                let cached_ranges: Vec<_> = col_cache.keys().cloned().collect();
+                trace!(
+                    "For row group col {:?}, Cached ranges: {:?}",
+                    row_group_col,
+                    cached_ranges
+                );
+
+                for cached_range in &cached_ranges {
+                    if cached_range.start >= col_range.end {
+                        break;
+                    }
+
+                    if let Some(intersection) = intersect_ranges(&cached_range, remaining_range) {
+                        let data = col_cache.get(cached_range).unwrap();
+                        let part_start = intersection.start;
+                        let part_end = intersection.end;
+                        let data_offset = part_start - cached_range.start;
+
+                        if part_start == remaining_range.start {
+                            let part_data =
+                                data.slice(data_offset as usize..(data_offset + (part_end - part_start)) as usize);
+                            let part = Part::new(id.clone(), part_start, part_data);
+
+                            trace!("Pushing part to queue from cache: {:?}", part_start..part_end);
+                            part_queue_producer.push(Ok(part));
+
+                            *remaining_range = part_end..remaining_range.end;
+                            if remaining_range.is_empty() {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    false
+}
+
+/// Fetches the requested range from the client
+#[allow(clippy::too_many_arguments)]
+async fn fetch_from_client<Client>(
+    client: &Client,
+    bucket: &str,
+    id: &ObjectId,
+    remaining_range: &mut Range<u64>,
+    preferred_part_size: usize,
+    in_mem_cache: &InMemoryCache,
+    cols: &mut Vec<((RowGroupIndex, ColumnIndex), Range<u64>)>,
+    part_queue_producer: &PartQueueProducer<Client::ClientError>,
+) -> Result<(), PrefetchReadError<Client::ClientError>>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    trace!("Fetching from client for range: {:?}", remaining_range);
+    let prefetch_range = RequestRange::new(
+        remaining_range.end as usize,
+        remaining_range.start,
+        cmp::min(preferred_part_size as u64, remaining_range.end - remaining_range.start) as usize,
+    );
+
+    match get_from_client(client, bucket, id, prefetch_range).await {
+        Ok(parts) => {
+            let mut rowgroup_col_cache = in_mem_cache.write().await;
+            rowgroup_col_cache.get_or_insert_with(HashMap::new);
```

Comment:
You get given the inner value here, you should use it so you can drop the unwrap later.

```suggestion
            let mut rowgroup_col_cache = in_mem_cache.write().await;
            let rowgroup_col_cache = rowgroup_col_cache.get_or_insert_with(HashMap::new);
```